src/System.Management.Automation/engine/Modules/PSModuleInfo.cs

```diff
@@ -577,6 +577,15 @@ private bool IsScriptModuleFile(string path)
             var res = new Dictionary<string, TypeDefinitionAst>(StringComparer.OrdinalIgnoreCase);
             foreach (var nestedModule in this.NestedModules)
             {
+                if (nestedModule == this)
+                {
+                    // Circular nested modules could happen with ill-organized module structure.
+                    // For example, module folder 'test' has two files: 'test.psd1' and 'test.psm1', and 'test.psd1' has the following content:
+                    //    "@{ ModuleVersion = '0.0.1'; RootModule = 'test'; NestedModules = @('test') }"
+                    // Then, 'Import-Module test.psd1 -PassThru' will return a ModuleInfo object with circular nested modules.
+                    continue;
```

Comment:
Based on Issue which we fix I believe it make sense to inform users about the situation or even terminated the cmdlet.

Reply:
Turning something that works to an error is a hard breaking change. As for the warning, I'm not sure if we should do it, since technically import-module is doing what it's asked to do ... (have itself as the nested module).

Reply:
Can ScriptAnalyzer catch this?

Reply:
Issue opened: https://github.com/PowerShell/PSScriptAnalyzer/issues/997